src/pattern-library/components/Library.tsx

```diff
@@ -265,6 +265,86 @@ function Demo({
   );
 }
 
+export type LibraryExampleDemoProps = Omit<LibraryDemoProps, 'children'> & {
+  /**
+   * Example file to read and use as content (to be rendered with syntax
+   * highlighting)
+   */
+  exampleFile: string;
+};
+
+function ExampleDemo({
```

Comment:
This would benefit from a comment to state what the component is for. "ExampleDemo" is not super descriptive. Did you plan to consolidate this with `Library.Demo` so there is just one component, or just share parts of the implementation?

Reply:
> Did you plan to consolidate this with `Library.Demo` so there is just one component, or just share parts of the implementation?

I haven't decided yet, but having a single component would be good if it doesn't make things too messy.

In an ideal world, we should aim to allow these three scenarios:

1. We pass explicit children to <Library.Demo />` and the source example is generated from it (the already supported use case):
    ```tsx
    <Library.Demo title="Basic Thumbnail" withSource>
      <div className="w-[250px]">
        <Thumbnail>
          <img src="https://placekitten.com/400/400" alt="kitty" />
        </Thumbnail>
      </div>
    </Library.Demo>
    ```
2. We pass an `exampleFile` prop and let both children and source be generated from it (what this PR is introducing)
    ```tsx
    <Library.Demo title="Basic Thumbnail" withSource exampleFile="basic-thumbnail" />
    ```
3. Any of the options above, but the source code is also explicitly provided the same way we do in `<Library.Code />`, just in case explicit children or the external example file do not fully cover the usage example we want to showcase. 
    ```tsx
    <Library.Demo 
      title="Basic Thumbnail" 
      withSource 
      exampleFile="basic-thumbnail"
      code={`
        import Foo from '...';

        function Foo() {
          // [...]
        }
      `}
    />
    ```
    ```tsx
    <Library.Demo 
      title="Basic Thumbnail" 
      withSource 
      code={`
        import Foo from '...';

        function Foo() {
          // [...]
        }
      `}
    >
      <div className="w-[250px]">
        <Thumbnail>
          <img src="https://placekitten.com/400/400" alt="kitty" />
        </Thumbnail>
      </div>
    </Library.Demo>
    ```

We can leave case 3 out of this PR for simplicity, but have it in mind as part of the design.

All those cases will require making some decisions in the component's props API, like not allowing to pass both children and `exampleFile`, or assuming `withSource` is true if `code` is provided.

Reply:
I have finally included the new behavior in the existing `Demo` component via a new `useDemoContentAndSource` hook which has the conditional logic.

With that, cases 1 and 2 from my comment above should be possible.